File: 144. 二叉树的前序遍历/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# 递归版本
# class Solution:
#     def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
#         if not root:
#             return []
#         left = self.preorderTraversal(root.left)
#         right = self.preorderTraversal(root.right)
#         return [root.val] + left + right

# 迭代版本
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack = []
        ans = []
        if root:
            stack.append(root)
        else:
            return []
        while stack:
            cur = stack.pop()
            ans.append(cur.val)
            # 模拟递归的出入栈，所以要右节点先入栈左节点后入栈才能保证先序遍历
            if cur.right:
                stack.append(cur.right)
            if cur.left:
                stack.append(cur.left)
        return ans

File: 144. 二叉树的前序遍历/Solution.cpp
#include<bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// 递归版本
// class Solution {
// public:
//     void traveral(TreeNode* root, vector<int>& ans){
//         if(root == nullptr){
//             return;
//         }
//         ans.push_back(root->val);
//         traveral(root->left, ans);
//         traveral(root->right, ans);
//     }
//     vector<int> preorderTraversal(TreeNode* root) {
//         vector<int> ans;
//         traveral(root, ans);
//         return ans;
//     }
// };

// 迭代版本
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        stack<TreeNode*> st;
        vector<int> ans;
        if(root) st.push(root);
        else return ans;
        while(!st.empty()){
            TreeNode* cur = st.top();
            st.pop();
            ans.push_back(cur->val);
            if(cur->right) st.push(cur->right);
            if(cur->left) st.push(cur->left);
        }
        return ans;
    }
};